compiler/passes/src/name_validation/program.rs

```diff
@@ -0,0 +1,79 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::NameValidationVisitor;
+
+use leo_ast::*;
+
+impl ProgramVisitor for NameValidationVisitor<'_> {
+    fn visit_program(&mut self, input: &Program) {
+        input.stubs.iter().for_each(|(_symbol, stub)| self.visit_stub(stub));
+        input.modules.values().for_each(|module| self.visit_module(module));
+        input.program_scopes.values().for_each(|scope| self.visit_program_scope(scope));
+    }
+
+    fn visit_program_scope(&mut self, input: &ProgramScope) {
+        let program_name = input.program_id.name;
+        self.check_name(program_name, "program", &[]);
+
+        input.structs.iter().for_each(|(_, function)| self.visit_struct(function));
+        input.functions.iter().for_each(|(_, function)| self.visit_function(function));
+    }
+
+    fn visit_module(&mut self, input: &Module) {
+        input.structs.iter().for_each(|(_, function)| self.visit_struct(function));
+        input.functions.iter().for_each(|(_, function)| self.visit_function(function));
+    }
+
+    fn visit_stub(&mut self, input: &Stub) {
+        input.structs.iter().for_each(|(_, function)| self.visit_struct_stub(function));
+        input.functions.iter().for_each(|(_, function)| self.visit_function_stub(function));
+    }
+
+    fn visit_struct(&mut self, input: &Composite) {
+        let struct_name = input.identifier;
+        let item_type = if input.is_record { "record" } else { "struct" };
+        self.check_name(struct_name, item_type, &[]);
+
+        for Member { identifier: member_name, .. } in &input.members {
+            if input.is_record {
+                self.check_name(*member_name, "record member", &["owner"]);
+            } else {
+                self.check_name(*member_name, "struct member", &[]);
+            }
+        }
+    }
+
+    fn visit_function(&mut self, function: &Function) {
+        use Variant::*;
+        match function.variant {
+            Transition | AsyncTransition => self.check_name(function.identifier, "transition", &[]),
+            Inline | Function | AsyncFunction | Script => {}
+        }
+    }
+
+    fn visit_function_stub(&mut self, input: &FunctionStub) {
+        use Variant::*;
+        match input.variant {
+            Transition | AsyncTransition => self.check_name(input.identifier, "transition", &[]),
+            Inline | Function | AsyncFunction | Script => {}
```

Comment:
Same here. Though stubs should really be validated before we get to this point, but a sanity check is good nonetheless.